### DISCUSSION TITLE
Errors with Emacs 30.1's visual-wrap-prefix-mode

### DISCUSSION THREAD

**Discussion by krisbalintona** [minad/org-modern#238] [Q&A]
Hi,

My version of Emacs is:

    GNU Emacs 31.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.43, cairo version 1.18.2) of 2024-10-20

Recently, `visual-wrap-prefix-mode` (the successor to `adaptive-wrap` on GNU-Elpa, as far as I can tell), was added to Emacs 30.1. It's meant to prefix lines visually; that is, visually wrapped lines are left-aligned with the beginning of the text it continues.

At least from my testing, when using org-modern with `visual-wrap-prefix-mode` enabled, errors like the following appear when I call e.g. `consult-line` (with `toggle-debug-on-error` enabled):

    Debugger entered--Lisp error: (wrong-type-argument listp #("▶" 0 1 (face org-modern-symbol)))
      add-display-text-property(0 2 min-width nil #("* " 0 1 (keymap (keymap (C-down-mouse-1 . org-mouse-move-tree-start) (C-drag-mouse-1 . org-mouse-move-tree) (follow-link . mouse-face) (mouse-3) (mouse-2 . org-open-at-mouse)) mouse-face highlight display #("▶" 0 1 (face org-modern-symbol)) face org-level-1 fontified t) 1 2 (keymap (keymap (C-down-mouse-1 . org-mouse-move-tree-start) (C-drag-mouse-1 . org-mouse-move-tree) (follow-link . mouse-face) (mouse-3) (mouse-2 . org-open-at-mouse)) mouse-face highlight face org-level-1 fontified t)))
      visual-wrap--content-prefix(#("* " 0 1 (keymap (keymap (C-down-mouse-1 . org-mouse-move-tree-start) (C-drag-mouse-1 . org-mouse-move-tree) (follow-link . mouse-face) (mouse-3) (mouse-2 . org-open-at-mouse)) mouse-face highlight display #("▶" 0 1 (face org-modern-symbol)) face org-level-1 fontified t) 1 2 (keymap (keymap (C-down-mouse-1 . org-mouse-move-tree-start) (C-drag-mouse-1 . org-mouse-move-tree) (follow-link . mouse-face) (mouse-3) (mouse-2 . org-open-at-mouse)) mouse-face highlight face org-level-1 fontified t)) 144)
      visual-wrap--apply-to-line(144)
      visual-wrap-prefix-function(1 32963)
      #f(compiled-function (fun) #<bytecode 0x1baa355fbb1815bf>)(visual-wrap-prefix-function)
      jit-lock--run-functions(1 32963)
      jit-lock-fontify-now()
      consult--fontify-all()
      consult--line-candidates(nil 882)
      #f(compiled-function () #<bytecode -0xca8f0d9a3e20413>)()
      consult-line(nil nil)
      funcall-interactively(consult-line nil nil)
      command-execute(consult-line)
    
Not sure whether org-modern's implementation, consult, `visual-wrap-prefix-mode` is at fault, but I thought I'd create an issue here first since I haven't run across any malfunctions with `visual-wrap-prefix-mode` beside this one.

**Comment by minad** [comment]
Hi, I cannot reproduce this problem. Do you have a minimal recipe and an example Org file which demonstrates the issue? It could be that `visual-wrap-prefix-mode` is not compatible with `org-modern-mode` because both use `wrap-prefix` and `line-prefix` text properties.

**Reply by krisbalintona** [reply]
> Hi, I cannot reproduce this problem. Do you have a minimal recipe and an example Org file which demonstrates the issue? It could be that `visual-wrap-prefix-mode` is not compatible with `org-modern-mode` because both use `wrap-prefix` and `line-prefix` text properties.

It was a bit of a headache but I think I made some progress. I don't know much about how font locking works, but I think it has something to do with unfontified regions of an org buffer. 

I was having trouble creating a small org file to reproduce the issue, but it was consistent with my large literate config. So I tried making a simple org file (headings + content) long enough so that it was longer than a screenful. For me, I just pasted the following around 40 times into an org buffer:
``` org
* Heading
Pellentesque dapibus suscipit ligula.  Donec posuere augue in quam.  Etiam vel tortor sodales tellus ultricies commodo.  Suspendisse potenti.  Aenean in sem ac leo mollis blandit.  Donec neque quam, dignissim in, mollis nec, sagittis eu, wisi.  Phasellus lacus.  Etiam laoreet quam sed arcu.  Phasellus at dui in ligula mollis ultricies.  Integer placerat tristique nisl.  Praesent augue.  Fusce commodo.  Vestibulum convallis, lorem a tempus semper, dui dui euismod elit, vitae placerat urna tortor vitae lacus.  Nullam libero mauris, consequat quis, varius et, dictum id, arcu.  Mauris mollis tincidunt felis.  Aliquam feugiat tellus ut neque.  Nulla facilisis, risus a rhoncus fermentum, tellus tellus lacinia purus, et dictum nunc justo sit amet elit.
```
And with the following recipe in an emacs -q (actually, a `package-isolate` with the org-modern and consult packages):
``` emacs-lisp
(progn
    (use-package org-modern :hook (org-mode . org-modern-mode))
    (add-hook 'org-mode-hook #'visual-wrap-prefix-mode)
    (define-key search-map "l" #'consult-line))
```
When I go into this long org file and call `consult-line` the error described in my initial post occurs, at least on the first invocation. However, if I kill the buffer and reopen the file (or just `revert-buffer-quick`), then scroll down so that I reach the bottom and at least every part of the buffer has been seen, then the error doesn't occur.

Here is my Emacs version:
```
GNU Emacs 31.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.43, cairo version 1.18.2) of 2024-10-20
```
and my org version:
```
Org mode version 9.7.13 (9.7.13-8566bc @ /home/krisbalintona/emacs-refresh/elpa/org-9.7.13/)
```
and I am on org-modern version 1.5.

Hope this helps.

**Reply by minad** [reply]
Still no success in reproducing this. I've listed visual-wrap-prefix-mode here for now: https://github.com/minad/org-modern?tab=readme-ov-file#incompatibilities

**Reply by krisbalintona** [reply]
> Still no success in reproducing this. I've listed visual-wrap-prefix-mode here for now: https://github.com/minad/org-modern?tab=readme-ov-file#incompatibilities

Hm, that's unfortunate. It's quite replicable on my end with emacs -q, so it might have to do with the combination of versions I have for Emacs, and org-modern, consult.

I might try debugging again with different versions. If I do, I'll report back here.
